### DISCUSSION TITLE
Question about the us_socket_context_t and HttpContext

### DISCUSSION THREAD

**Discussion by trunglebka** [uNetworking/uWebSockets#1447] [Q&A]
Hi, I'm new to C++ without any knowledge of C (comming from Java/Netty). I'm currently investigate to this project to learn some development skill and have a question that need help.
From EchoServer example, I see that we need uWS::App to use this lib. Tracing down to TemplatedApp I see that you use casting "(struct us_socket_context_t *) httpContext" extensively but I don't see any common between "us_socket_context_t" and "HttpContext<SSL>" so I don't understand how it can work.

Can you hint/help me something so that I can investigate more? Please...

**Comment by ghost** [comment]
Yes that's called black magic 😘

**Reply by trunglebka** [reply]
Yeah, with me it is truly "magical".
Can you help me to understand this? May be some article or keywords... :'(

**Comment by ghost** [comment]
I think "pointer black magic" is kind of the best description here. You don't need explicit hierarchies if things map properly either way. But this is not relating to uWS

**Reply by trunglebka** [reply]
Yeah I know that this is not relating to uWS, I just call for your help to make the life easy for me - an self taught C++ amateur and trying to learn the way you make this lib both easy to use interface and high performance at it core (after some searching, of course). 

I understand that we can cast pointer to any kind as long as thing can map properly. However, from what I have read in this lib I don't understand how thing can align together.
Let specific:
(1)
```
HttpContext<SSL> *init() {
        /* Handle socket connections */
        us_socket_context_on_open(SSL, getSocketContext(), [](us_socket_t *s, int /*is_client*/, char */*ip*/, int /*ip_length*/) {});
//...
}
```

(2)
```
void us_socket_context_on_open(int ssl, struct us_socket_context_t *context, struct us_socket_t *(*on_open)(struct us_socket_t *s, int is_client, char *ip, int ip_length)) {
//...
    context->on_open = on_open;
}
```

In (1): getSocketContext() is "us_socket_context_t" cast of "HttpContext"
In (2):  context->on_open = on_open; 

I can't see what element of HttpContext corresponding to us_socket_context_t::on_open, so I can't understand how the real object (HttpContext) can save "on_open" so it can work.
Please help me if you have time. Thank you.

**Comment by ghost** [comment]
uWebSockets is built on uSockets.

uSockets is an **opaque** C library. Meaning, its exact type definitions are **unknown** (opaque is the opposite of transparent) for anything using it. For all uWebSockets knows, anything provided by uSockets is just a simple undefined **void pointer**. That's why we do these very liberal casts between seemingly unrelated types.

It works because no function has any idea about, or assumes, how the actual implementation of uSockets types look like.

The hierarchy is like so:

* us_socket_context_t is some kind of "context" that holds sockets and knows about callbacks and shared settings, etc
* us_socket_t is some kind of socket, that relates to some us_socket_context_t
* The memory layout of all types are unknown, but there is user data stored as us_socket_ext and us_socket_context_ext
* HttpContext is castable to us_socket_context_t BUT CANNOT HAVE A SIZE (because we do not know the size, opaque)
* HttpContextData is the thing living in us_socket_context_ext AND DOES HAVE A SIZE
* WebSocket is castable to us_socket_t BUT CANNOT HAVE A SIZE
* WebSocketData is the thing living in us_socket_ext AND DOES HAVE A SIZE
* HttpResponse is castable to us_socket_t BUT CANNOT HAVE A SIZE
* HttpResponseData is the thing living in us_socket_ext AND DOES HAVE A SIZE
* etc
* etc

If you look at HttpContext https://github.com/uNetworking/uWebSockets/blob/master/src/HttpContext.h#L37 you can see that no function ever dereferences "this" - it always treats "this" as a void pointer without a size. All it does is pass "this" as if it indeed was us_socket_context_t, which it is. So the cast to HttpContext never really does anything at all. It's just a perspective change.

**Reply by trunglebka** [reply]
Thank you for the detailed information.
